### PR TITLE
feat(thanos): adding store and tests

### PR DESCRIPTION
**Pull request by viennaa**
adding core Thanos Store functionality along with some changes
* default mechanism for sidecar endpoint changed to be templated instead of configured in the plugindefinition in order to enable standalone functionality
* standalone functionality added, to have an empty querier without the default stores
* documentation added
* initContainer for sidecar mount needs to run as root to set respective permissions, added for Store with emptydir too
* compactor annotations directive fixed

### CODE REVIEW

File: thanos/charts/templates/query.yaml [hunk 1]
```diff
@@ -53,9 +53,15 @@ spec:
         {{- range .Values.thanos.query.additionalArgs }}
         - {{ . }}
         {{- end }}
+        {{/* Technically it is possible to add stores and have the default stores all together.*/}}
+        {{/* Enabling standalone would remove the default stores for an empty store list to be maintained entirely manual.*/}}
```
Comment (richardtief) [inline thanos/charts/templates/query.yaml:57]: As discussed: This message should also be forwarded to the user. Additionally fail or warn if no store is set.

File: thanos/plugindefinition.yaml [hunk 2]
```diff
@@ -64,9 +64,13 @@ spec:
...
-        - default: 
-          - prometheus-operated:10901
-          description: Standard port to be exposed by Prometheus Operator. If multiple Prometheis are deployed, this needs to be switched to prometheus' CRD name suffixed with -prometheus.
+        - default: false
+          description: Removes standard Thanos endpoints for overarching Queries 
+          name: thanos.query.standalone
+          required: false
+          type: "bool"
+        - default: [] 
+          description: "Automatically filled if Query is not standalone. If multiple Prometheis are deployed, all Prometheus services and Thanos store services need to be added to this list manually. Usually this means Prometheus'-CRD name suffixed with -prometheus and Thanos' release name suffixed with -store."
```
Comment (richardtief) [inline thanos/plugindefinition.yaml:73]: I would really like the required services to be added automatically in the future. 
```suggestion
          description: "Automatically filled if Query is not standalone. If multiple Prometheis are deployed, all Prometheus, Thanos Query and Thanos Store services need to be added to this list manually. e.g. kube-monitoring-prometheus:9090 or thanos-store:10901"
```

File: thanos/plugindefinition.yaml [hunk 3]
```diff
@@ -64,9 +64,13 @@ spec:
...
           required: false
           type: "map"
-        - default: 
-          - prometheus-operated:10901
-          description: Standard port to be exposed by Prometheus Operator. If multiple Prometheis are deployed, this needs to be switched to prometheus' CRD name suffixed with -prometheus.
+        - default: false
+          description: Removes standard Thanos endpoints for overarching Queries 
+          name: thanos.query.standalone
+          required: false
+          type: "bool"
```
Comment (richardtief) [inline thanos/plugindefinition.yaml:71]: NIT: The `" "` around the type are not needed although it's a string. Applies to all fields.
```suggestion
          type: bool
```

**Review by richardtief — CHANGES_REQUESTED** [review]
Cool! Just some comments below.

File: thanos/README.md [hunk 4]
```diff
@@ -81,22 +81,58 @@ spec:
...
 
 `--store=prometheus-operated:10901`
 
-This would be fine, unless you've got more than one prometheus running in this namespace. Then you would need to add all prometheus stores dedicatedly:
+### Thanos Endpoint
+Thanos Query would check for a Thanos endpoint named like `releaseName-store`. In the cli this param would look like:
+
+`--store=thanos-kube-store:10901`
+
+If you just have a single Thanos the default option would work and does not need anything else.
```
[the remark this answers is not in the document]
Comment (viennaa) [reply thanos/README.md:99]: Rather a single thanos plugin deployed, I'ma refine

File: thanos/plugindefinition.yaml (hunk 2 above)
Comment (viennaa) [reply thanos/plugindefinition.yaml:73]: I suppose this was outputted by the greenhousectl tooling, but granted.

Comment (viennaa) [reply thanos/plugindefinition.yaml:73]: Well yes, we've got to think about this.

**Comment by viennaa** [comment]
Basic testing added

File: thanos/charts/templates/tests/test-thanos-config.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        verify "there is 1 service named '{{ .Release.Name }}-query'"
+        verify "there is 1 service named '{{ .Release.Name }}-store'"
+        try "at most 2 times every 5s to get pods named '{{ .Release.Name }}-compactor' and verify that '.status.phase' is 'running'"
+        try "at most 4 times every 5s to get pods named '{{ .Release.Name }}-query' and verify that '.status.phase' is 'running'"
+        try "at most 4 times every 5s to get pods named '{{ .Release.Name }}-store' and verify that '.status.phase' is 'running'"
+    }
+
+    @test "Verify successful creation and bound status of {{ .Release.Name }} persistent volume claims" {
+        try "at most 3 times every 5s to get persistentvolumeclaims named '{{ .Release.Name }}-compactor' and verify that '.status.phase' is 'Bound'"
+    }
```
Comment (ibakshay) [inline thanos/charts/templates/tests/test-thanos-config.yaml:46]: I really like the tests. 🚀

File: thanos/charts/templates/tests/test-thanos-config.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    #!/usr/bin/env bats
+# SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+# SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+# SPDX-License-Identifier: Apache-2.0
+
+# SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+# SPDX-License-Identifier: Apache-2.0
```
Comment (ibakshay) [inline thanos/charts/templates/tests/test-thanos-config.yaml:24]: Do we need the license information here? 
If yes, then shall we remove the redundant lines? 

```suggestion
# SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
# SPDX-License-Identifier: Apache-2.0
```

Comment (viennaa) [reply thanos/charts/templates/tests/test-thanos-config.yaml:24]: Not sure how this slipped through, 👍

**Comment by viennaa** [comment]
Bwhahaha the licenses keep coming in :D

**Review by richardtief — APPROVED** [review]

**Pull request merged**